### PR TITLE
Speed up ScubaGear - Add execution tracing - Fix multiple bugs

### LINKED ISSUE #1313: Add timing information to long running ScubaGear functions
# 💡 Summary #

ScubaGear generates a set of results in several formats.  The idea is to add timing information for the assessment run associated with the results to include timing information about the wall-clock execution time of the run, and possibly for each major product and component.  

## Motivation and context ##

Doing so can better inform users of where ScubaGear may be spending most of its time assessing, and provide feedback to developers to help optimize performance.

## Implementation notes ##

Implementing timing information for the run could be implemented by:

- Defining JSON fields for timing information to include in the ScubaResults output
- Adding timing/profiling commands to ScubaGear execution at the orchestrator and provider levels

## Acceptance criteria ##

How do we know when this work is done?

- [x]  The ScubaResults.json includes a section for timing information that successfully tracks either start/stop times or execution duration for the overall program
- [x] ScubaResults.json includes timing information for each provider included in the assessment run
- [x] ScubaResults.json includes timing information for major phases of ScubaGear execution, including providers (aggregate), Rego test execution, and report/output creation

### LINKED ISSUE #2086: ScubaGear 1.7.1 has delayed module import issues.
### Prerequisites

- [x] This issue has an informative and human-readable title.

### ScubaGear Version

1.7.1

### Operating System

Windows 11

### PowerShell Version

5.1

### M365 Environment and License(s)

All environments

### 🐛 Summary

In ScubaGear , module import is delayed due to duplicated initialization logic executed across multiple scripts. Specifically:

- `dependencies.ps1` and `CheckVersion.ps1` both perform overlapping module/version validation
- Running `Import-Module ScubaGear` triggers **two separate version checks**
- `Test-ScubaGearVersion` introduces redundant logic instead of being the single source of truth

This results in:
- Increased module load time  
- Redundant processing  
- Confusing and inconsistent output to the user 

### Steps to reproduce

1. Install or update to ScubaGear v1.7.1  
2. Run:
   ```powershell
   Import-Module ScubaGear -Verbose
3.   Observe:
  - Multiple version check executions
  - Repeated or unclear output messages related to version validation

### Expected behavior

Module import should:

- Execute a single, centralized version check
- Avoid duplicate logic across scripts
- Provide clear, concise output indicating:
  - Current version
  - Required/expected version (if applicable)
  - Any actionable remediation steps

### Output from Initialize-SCuBA (optional)

_No response_

### PR DESCRIPTION
**Pull request by tkol2022**
## 🗣 Description ##

- Makes ScubaGear startup considerably faster be removing the logging of all the PowerShell modules installed on the system which we don't need when troubleshooting. Since ScubaGear starts faster it makes the Functional Tests in GitHub execute about 70% faster. This also helps developers be more efficient since we lose at lot of time waiting for functional tests to complete. Functional tests are a critical component of every developer's process.
- Adds Trace messages with functional call execution times to each Provider which helps users get feedback on how the tool is progressing instead of having them just wait for ScubaGear to finish. It also helps our dev team track how the tool is executing and pinpoint areas of performance problems or other execution troubleshooting, including in GitHub Actions logs.
- Fixes the total_user_count value in the ScubaResults JSON which was getting set to -1 in every tenant. I didn't bother creating an issue for this so as not to flood our issues list.
- Fixes duplicate checks for PowerShell dependency modules when ScubaGear is imported, thereby reducing import time.
- Adds a new developer-only environment variable named $env:DEV_SKIP_DEPENDENCY_CHECK that bypasses checking the PowerShell module dependencies on developer machines. Since we import ScubaGear a lot, this saves considerable time each day. No impact to the end user or to GitHub Actions.
- Moves the slower running functions that retrieve privileged users and risky applications to the bottom of ExportAADProvider and provides a helpful message to the user that those will take longer. I added a helpful message to the user inside ExportSecuritySuiteProvider when it connects to the Security and Compliance REST API since that takes long.
- I changed some Unicode characters to ASCII inside ScubaLogging.psm1 since they were causing execution problems due to file encoding in edge cases.
- I fixed a couple of PowerShell unit tests that had missing mocked functions and could lead to unstable results.
- I removed the unused function UpdateDirectorySettingByName from FunctionalTestUtils.ps1. This is a legacy file leftover from a previous aeon of our universe.

closes #1313
closes #2086

Will help diagnose the performance issue observed in the publicly reported issue #1718

## Quick visual
This is what the enhanced ScubaGear output looks like:

image

## 🧪 Testing Guidance ##

- This is one of those "you should test against all 4 tenants" PR.
- Run the functional tests against any of the products and verify that they run faster compared to main.
- Validate that all of the providers produce trace messages to the console and to the log file. For AAD, validate that the execution executes the faster functions first and then alerts the user in the console that the slower functions (like computing the risky applications) are about to commence.
- Verify that the total_user_count value in the ScubaResults JSON is accurate based on how many users are in the respective tenant.

### To test the new developer environment variable which makes ScubaGear import faster

First determine how long it takes ScubaGear to import against the main branch:

```
Measure-Command { Import-Module .\PowerShell\ScubaGear }
```

image

Then determine how long it takes against this PR branch with the environment variable:

```
$env:DEV_SKIP_DEPENDENCY_CHECK = 1

Measure-Command { Import-Module .\PowerShell\ScubaGear }
```

image

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] Changes are sized such that they do not touch excessive number of files.
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [x] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] All relevant type-of-change labels added.
- [x] All relevant project fields are set.
- [x] All relevant repo and/or project documentation updated to reflect these changes.
- [x] Unit tests added/updated to cover PowerShell and Rego changes.
- [x] Functional tests added/updated to cover PowerShell and Rego changes.
- [x] All relevant functional tests passed.
- [x] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

- [x] PR passed smoke test check.
- [x] Feature branch has been rebased against changes from parent branch, as needed.

  Use `Update branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [x] Resolved all merge conflicts on branch.
- [x] Squash all commits into one PR level commit using the `Squash and merge` button.

## ✅ Post-merge checklist ##

- [x] Feature branch deleted after merge to clean up repository.
- [x] Close issues resolved by this PR if the [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) did not activate.
- [x] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

### CODE REVIEW

File: PowerShell/ScubaGear/Modules/Providers/ExportAADProvider.psm1 [hunk 1]
```diff
@@ -58,17 +58,78 @@ function Export-AADProvider {
...
+    # dedicated hybrid application, or they are still in the legacy hybrid configuration.
+    $LegacyExchangeSP =  ConvertTo-Json -Depth 4 @(
+        $Tracker.TryCommand("Get-LegacyExchangeServicePrincipal", @{
+            "M365Environment"=$M365Environment
+        })
+    )
+
+    $DedicatedExchangeHybridApps = ConvertTo-Json -Depth 4 @(
+        $Tracker.TryCommand("Get-DedicatedExchangeHybridApplications", @{
+            "AggregateRiskyAppsRaw"=$AggregateRiskyAppsRaw
```
Comment (mitchelbaker-cisa) [inline PowerShell/ScubaGear/Modules/Providers/ExportAADProvider.psm1]: `$DedicatedExchangeHybridApps` relies on the output from `$AggregateRiskyAppsRaw` to determine if an exchange hybrid app is configured. This refactor causes the output of `$DedicatedExchangeHybridApps` to be null since `$AggregateRiskyAppsRaw` is referenced before the variable is initialized.

Below is the output for `dedicated_exchange_hybrid_applications` for Tenant_1 (functional test code name) from this feature branch:

Image

Below is the output for `dedicated_exchange_hybrid_applications` for Tenant_1 (functional test code name) from main:

Image

See the invocation order below:

Image

File: PowerShell/ScubaGear/CheckVersion.ps1 [hunk 2]
```diff
@@ -1,32 +0,0 @@
-
-function Invoke-CheckScubaGearVersion {
```
Comment (mitchelbaker-cisa) [inline PowerShell/ScubaGear/CheckVersion.ps1:2]: This function was a feature Mike added to output if a new version of ScubaGear was available. This was broken for v1.7.1 because I made the mistake of not bumping the module version but should be resolved with v1.8.0. This should be fine to leave in the tool unless it has some overlap with the ScubaLogging dependency check logic you removed to increase the tool performance?

**Review by mitchelbaker-cisa — CHANGES_REQUESTED** [review]
Testing steps completed:

- ran feature branch against all tenants and confirmed `total_user_count` is the same as the total number of users in each tenant.
- performed a side-by-side comparison of functional tests run between main vs this branch:

| Defender Tenant        | Main Time   | Feature Branch Time |
|----------------|-------------|---------------------|
| Def_Tenant1    | 17m 51s     | 6m 40s             |
| Def_Tenant2    | 16m 3s      | 3m 50s             |
| Def_Tenant6    | 14m 44s     | 3m 57s             |
| Def_Tenant7    | 15m 53s     | 4m 31s             |
| Def_Tenant1_G5 | 3m 1s       | 2m 57s             |
| Def_Tenant2_G3 | 5m 34s      | 2m 33s             |
| Def_Tenant6_G5 | 2m 26s      | 1m 38s             |
| Def_Tenant7_G5 | 2m 43s      | 1m 39s             |

| Entra Tenant         | Main Time   | Feature Branch Time |
|----------------|-------------|---------------------|
| Ent_Tenant1    | 39m 10s     | 11m 46s            |
| Ent_Tenant2    | 38m 35s     | 9m 54s             |
| Ent_Tenant6    | 36m 52s     | 10m 10s            |
| Ent_Tenant7    | 38m 57s     | 9m 48s             |
| Ent_Tenant1_G5 | 15m 31s     | 6m 20s             |
| Ent_Tenant2_G3 | 6m 30s      | 2m 43s             |
| Ent_Tenant6_G5 | 16m 42s     | 5m 35s             |
| Ent_Tenant7_G5 | 16m 23s     | 5m 37s             |

| Exchange Tenant         | Main Time   | Feature Branch Time |
|----------------|-------------|---------------------|
| Exc_Tenant1    | 16m 48s     | 6m 19s             |
| Exc_Tenant2    | 16m 53s     | 5m 49s             |
| Exc_Tenant6    | 13m 46s     | 4m 22s             |
| Exc_Tenant7    | 16m 27s     | 6m 9s              |

| Power Platform Tenant           | Main Time   | Feature Branch Time |
|------------------|-------------|---------------------|
| Pwr_Tenant1      | 14m 23s     | 2m 52s             |
| Pwr_Tenant2      | 14m 34s     | 2m 53s             |
| Pwr_Tenant6      | 14m 36s     | 3m 5s              |
| Pwr_Tenant7      | 14m 47s     | 2m 58s             |

| SharePoint Tenant           | Main Time   | Feature Branch Time |
|------------------|-------------|---------------------|
| Shr_Tenant1_PNP  | 21m 27s     | 4m 42s             |
| Shr_Tenant2_PNP  | 21m 51s     | 4m 11s             |
| Shr_Tenant6_PNP  | 22m 58s     | 4m 10s             |
| Shr_Tenant7_PNP  | 22m 3s      | 4m 37s             |

| Teams Tenant           | Main Time   | Feature Branch Time |
|------------------|-------------|---------------------|
| Tms_Tenant1      | 25m 20s     | 6m 46s             |
| Tms_Tenant2      | 22m 11s     | 5m 30s             |
| Tms_Tenant6      | 21m 25s     | 4m 16s             |
| Tms_Tenant7      | 21m 13s     | 5m 37s             |
| Tms_Tenant1_GCC  | 3m 54s      | 1m 30s             |
| Tms_Tenant2_GCC  | 3m 23s      | 1m 29s             |
| Tms_Tenant6_GCC  | 2m 58s      | 1m 46s             |
| Tms_Tenant7_EX   | 5m 35s      | 2m 27s             |

Based on the output above, you can clearly see a huge performance lift for each tenant. Great work! This cuts down the functional test execution time from about ~55m on average down to ~25m. 

- Noticed a slight performance improvement when importing the ScubaGear module. I tested this by opening a new terminal window then running the Measure-Command function.

main:

Image

feature branch:

Image

**Review by dagarwal-ecs — CHANGES_REQUESTED** [review]
Nevermind fixed the issue I was having

**Review by dagarwal-ecs — APPROVED** [review]
Ran against all tenants, and it worked well and checked the total user count to ensure it was correct

File: PowerShell/ScubaGear/CheckVersion.ps1 (hunk 2 above)
Comment (tkol2022) [reply PowerShell/ScubaGear/CheckVersion.ps1:2]: I resolved this by checking that Test-ScubaGearVersion which is called by Dependencies.ps1 when ScubaGear loads, does in fact check the scubagear version and is the authoritative cmdlet in our official documentation for end users.

The only thing missing from Test-ScubaGearVersion was a check for the command line switch below which I am adding. Once that is there, we are safe to remove CheckVersion.ps1 since it is in fact redundant.

```
if (-not [string]::IsNullOrWhiteSpace($env:SCUBAGEAR_SKIP_VERSION_CHECK)) {
            return
        }
```

File: PowerShell/ScubaGear/Modules/Providers/ExportAADProvider.psm1 (hunk 1 above)
Comment (tkol2022) [reply PowerShell/ScubaGear/Modules/Providers/ExportAADProvider.psm1]: I fixed this and just tested it to ensure it creates the JSON node.

image

Comment (mitchelbaker-cisa) [reply PowerShell/ScubaGear/Modules/Providers/ExportAADProvider.psm1]: @tkol2022 Thanks, I did a quick test run myself and seeing the data now.

**Review by mitchelbaker-cisa — APPROVED** [review]
All of my comments are resolved, thanks.
